src/routes.rs

```diff
@@ -454,41 +457,84 @@ fn post_blob_upload(
     optimisation, but is arguably less flexible.
     */
 
-    let rn = RepoName(repo_name);
-    let r = ci.request_upload(&rn);
-    Runtime::new().unwrap().block_on(r).map_err(|e| {
+    let repo_name = RepoName(repo_name);
+    let mut rt = Runtime::new().unwrap();
+
+    let req = ci.request_upload(&repo_name);
+
+    let up_info = rt.block_on(req).map_err(|e| {
         warn!("Error getting ref from backend: {}", e);
         Error::InternalError
-    })
+    })?;
+    
+    if let Some(mut digest) = uri.query()  {
+        if digest.starts_with("digest=") {
+            digest = &digest["digest=".len()..];
+            println!("Doing post with data");
+            let sink_f = ci.get_write_sink_for_upload(&repo_name, &up_info.uuid());
+            let sink = rt.block_on(sink_f);
+        
+            return match sink {
+                Ok(mut sink) => {
+                    // We have a monolithic upload
+                    let len = data.stream_to(&mut sink);
+                    match len {
+                        Ok(len) => {
+                            let digest = Digest(digest.to_string());
+                            let r = ci.complete_upload(&repo_name, &up_info.uuid(), &digest, len);
+                            rt.block_on(r).map_err(|_| Error::InternalError).map(|au| Upload::Accepted(au))
+                        }
+                        Err(_) => Err(Error::InternalError),
+                    }
+                }
+                Err(_) => {
+                    // TODO: this conflates rpc errors with uuid not existing
+                    // TODO: pipe breaks if we don't accept the whole file
+                    // Possibly makes us prone to DOS attack?
+                    warn!("Uuid {} does not exist, piping to /dev/null", &up_info.uuid());
+                    let _ = data.stream_to_file("/dev/null");
```

Comment:
can't we just break the connection instead?

Reply:
Not sure, I've not really looked into this tbh

Reply:
I think this is best left for another PR - do you want to investigate?

Reply:
I will create an issue for it, meanwhile I will approve